### PR TITLE
[ADD][11.0] Adding module project_agile

### PR DESCRIPTION
**Pull request by pnajman-modoolar**
Hello,

Module ``project_agile`` provides core framework for development of the agile methodologies like kanban, scrum, scrumban, etc.

This module depends on following custom modules:

- [project_key](https://github.com/OCA/project/pull/366)
- [project_workflow](https://github.com/OCA/project/pull/371)
- [project_portal](https://github.com/OCA/project/pull/364)

Kind Regards,
Petar

### CODE REVIEW

File: project_agile/README.rst [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+   :alt: Odoo Community Association
+   :target: https://odoo-community.org
+
+This module is maintained by the OCA.
+
+OCA, or the Odoo Community Association, is a nonprofit organization whose
+mission is to support the collaborative development of Odoo features and
+promote its widespread use.
+
+To contribute to this module, please visit http://odoo-community.org.
```
Comment (elicoidal) [inline project_agile/README.rst]: please see https://github.com/OCA/web/pull/881#pullrequestreview-99094143

File: project_agile/data/project_project.xml [hunk 2]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+# Copyright 2017 - 2018 Modoolar <info@modoolar.com>
+# License LGPLv3.0 or later (https://www.gnu.org/licenses/lgpl-3.0.en.html).
+-->
+<odoo>
+    <data noupdate="1">
```
Comment (elicoidal) [inline project_agile/data/project_project.xml]: directly in odoo tag as of v10

File: project_agile/data/project_task.xml [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+
+        <record id="project_task_moscow_priority_could_have" model="project.task.priority">
+            <field name="name">Could have</field>
+            <field name="description">
+                <![CDATA[
+<strong>‘Could Haves‘</strong> are features that are nice to have and could potentially be included without incurring too much effort or cost.
+These will be the first features to be removed from scope if the project’s timescales are later at risk.
+                ]]>
+            </field>
+            <!--<field name="order" eval="8"/>-->
```
Comment (elicoidal) [inline project_agile/data/project_task.xml]: remove dead code

File: project_agile/import/project_workflow.xml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+# Copyright 2017 - 2018 Modoolar <info@modoolar.com>
+# License LGPLv3.0 or later (https://www.gnu.org/licenses/lgpl-3.0.en.html).
+-->
+<project-workflow description="&lt;p&gt;Default Workflow.&lt;/p&gt;" name="Default Workflow" default-state="ToDo">
```
Comment (elicoidal) [inline project_agile/import/project_workflow.xml]: Could you format the xml for readability?

File: project_agile/models/base.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        string='Sequence',
+        default=10,
+    )
+
+    @api.multi
+    def unlink(self):
+        for item in self:
+            if item.system:
+                raise exceptions.ValidationError(_(
+                    "%s '%s' is a system record!\n"
```
Comment (elicoidal) [inline project_agile/models/base.py:44]: please remove "!" from error message

File: project_agile/models/mixins.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
+# Copyright 2017 - 2018 Modoolar <info@modoolar.com>
+# License LGPLv3.0 or later (https://www.gnu.org/licenses/lgpl-3.0.en.html).
+
+import logging
+from odoo import models, api
+
+_logger = logging.getLogger(__name__)
+
+
+class IdSearchMixin(models.AbstractModel):
```
Comment (elicoidal) [inline project_agile/models/mixins.py:10]: IMO should be in separate module

File: project_agile/models/project_agile_board.py [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+
+    workflow_stage_ids = fields.Many2many(
+        comodel_name='project.task.type',
+        related='workflow_id.stage_ids',
+        readonly=True
+    )
+
+    _sql_constraints = [
+        ('stage_unique', 'UNIQUE(board_id, state_id)',
+         "Column state must be unique per board!"),
```
Comment (elicoidal) [inline project_agile/models/project_agile_board.py:358]: Remove "!"

File: project_agile/models/project_project.py [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+    default_task_type_id = fields.Many2one(
+        comodel_name='project.task.type2',
+        related='type_id.default_task_type_id',
+        string='Default Task Type',
+        readonly=True,
+    )
+
+    agile_enabled = fields.Boolean(
+        string='Use Agile',
+        help='If checked project will be enabled for agile management',
```
Comment (elicoidal) [inline project_agile/models/project_project.py]: agile -> Agile

File: project_agile/models/project_task.py [hunk 9]
```diff
@@ -0,0 +1,691 @@
...
+            name=name, args=args, operator=operator, limit=limit
+        )
+
+
+class Task(models.Model):
+    _name = 'project.task'
+    _inherit = ['project.task', 'project.agile.mixin.id_search']
+
+    def _default_agile_order(self):
+        self.env.cr.execute("SELECT MAX(agile_order) + 1 FROM project_task")
```
Comment (elicoidal) [inline project_agile/models/project_task.py]: any reason to use SQL instead of standard ORM?

File: project_agile/models/mixins.py (hunk 6 above)
Comment (elicoidal) [inline project_agile/models/mixins.py:10]: IMO should be in separate module

**Review by elicoidal — CHANGES_REQUESTED** [review]
Module is huge, would it make sense to split some of it in different module?
(I know that a huge work has already been done and I am fine with current result)

File: project_agile/views/project_agile.xml [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+    <template id="portal_my_tasks_task_edit" inherit_id="project.portal_my_tasks" name="portal_my_tasks_task_edit">
+        <xpath expr="//h3" position="inside">
+            <button name='new_task' data-toggle="modal"
+                    data-target=".modal_new_task" class="btn btn-default pull-right mr8"
+                    groups="project_agile.group_project_portal_task_edit">
+                <i class="fa fa-plus fa-lg"/>
+            </button>
+        </xpath>
+
+        <xpath expr="//div[@t-if='tasks']//thead//th[hasclass('col-md-10')]" position="after">
```
Comment (LoisRForgeFlow) [inline project_agile/views/project_agile.xml]: This has been changed upstream: https://github.com/odoo/odoo/commit/dea07fc29116c36c8a3773eec0fce526a972ba08#diff-2a064778276c73499a520e03c6c99f7fL138

File: project_agile/views/project_project_views.xml [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+                <attribute name="name">open_tasks</attribute>
+                <attribute name="type">object</attribute>
+            </xpath>
+
+            <xpath expr="//t[@t-esc='record.task_count.value']/../../.." position="attributes">
+                <attribute name="name">open_tasks</attribute>
+                <attribute name="type">object</attribute>
+            </xpath>
+
+            <xpath expr="//a[@t-if='record.task_needaction_count.raw_value']" position="attributes">
```
Comment (LoisRForgeFlow) [inline project_agile/views/project_project_views.xml]: This has changed as well: https://github.com/odoo/odoo/commit/8438470c3a578a2b0f4f4187936466a754e601d7#diff-5a45d5b004d555a5805f0892644b41b7L332

**Review by LoisRForgeFlow — COMMENTED** [review]
Hello,

First of all thanks for the great and hard work here!

Just wanted to bring up to the discussion one issue: this modules has a lot of complex xpath expressions and they can easily make the module uninstallable as odoo updates the views upstream. I'm pointing in line two current occurrences.

Although, I don't know which would be the appropriate way to handle this and not to merge a module that can end being uninstallable after some time.

Thanks again for contributing this to OCA :smile:

File: project_agile/models/mixins.py (hunk 6 above)
Comment (pnajman-modoolar) [reply project_agile/models/mixins.py:10]: @elicoidal I agree. I'll extract the code into separate module but please tell me first, in which repository should this module go into?

File: project_agile/models/project_task.py (hunk 9 above)
Comment (pnajman-modoolar) [reply project_agile/models/project_task.py]: I didn't want to fetch complete task just to read agile_order..

File: project_agile/views/project_agile.xml (hunk 10 above)
Comment (pnajman-modoolar) [reply project_agile/views/project_agile.xml]: Hi @lreficent, Yes, I'm aware of this. For now, I'll remove everything related to portal.

File: project_agile/models/project_task.py (hunk 9 above)
Comment (elicoidal) [reply project_agile/models/project_task.py]: Still: I would suggest if possible to use ORM functions

Comment (pnajman-modoolar) [reply project_agile/models/project_task.py]: Fixed

File: project_agile/README.rst (hunk 1 above)
Comment (pnajman-modoolar) [reply project_agile/README.rst]: Fixed

File: project_agile/data/project_project.xml (hunk 2 above)
Comment (pnajman-modoolar) [reply project_agile/data/project_project.xml]: Fixed

File: project_agile/data/project_task.xml (hunk 3 above)
Comment (pnajman-modoolar) [reply project_agile/data/project_task.xml]: Removed ...

File: project_agile/models/base.py (hunk 5 above)
Comment (pnajman-modoolar) [reply project_agile/models/base.py:44]: Removed ...

File: project_agile/models/project_agile_board.py (hunk 7 above)
Comment (pnajman-modoolar) [reply project_agile/models/project_agile_board.py:358]: Removed ...

File: project_agile/models/project_project.py (hunk 8 above)
Comment (pnajman-modoolar) [reply project_agile/models/project_project.py]: Fixed

File: project_agile/models/mixins.py (hunk 6 above)
Comment (pnajman-modoolar) [reply project_agile/models/mixins.py:10]: I agree. I'll move into separate module but for which repository should I create pull request?
Do we leave it here in this repo or maybe you have a suggestion?

**Comment by metaminux** [comment]
Hello,

Looking at travis CI logs, it seems VERSION was not updated to 11.0 in .travis.yml

Checks are failing because of the dependency on the "project_key" module which is present in the OCA/project repository. Shouldn't we add this repo in the install section of .travis.yml ?

**Pull request closed without merging**